src/lib/Microsoft.Health.Fhir.Ingest.Template/CalculatedFunction/CachingExpressionEvaluatorFactory.cs

```diff
@@ -0,0 +1,41 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+using EnsureThat;
+
+namespace Microsoft.Health.Fhir.Ingest.Template.CalculatedFunction
+{
+    public class CachingExpressionEvaluatorFactory : IExpressionEvaluatorFactory
+    {
+        private IExpressionEvaluatorFactory _wrappedExpressionEvaluatorFactory;
+        private IDictionary<string, IExpressionEvaluator> _expressionCache;
```

Comment:
I believe you're correct, and that I need to update this to be thread safe. Each `EventData` is currently processed in a separate thread, so this class could called from multiple threads. @dustinburson and I had discussed adjusting that logic within the `MeasurementEventNormalizationService`, but I feel we should leave it unless there is data showing that otherwise.

Reply:
Just a note on thread safety, the IDictionary should be fine to be accessed from multiple threads.  What is not thread safe is concurrent gets with writes.  So if you create the dictionary and after it is created it is immutable you should be fine.  That said we should enforce the behavior contractually using IReadonlyDictionary over a normal IDictionary https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlydictionary-2?view=net-5.0

This would be preferable to a concurrent dictionary since it would avoid the locking overhead on reads when don't need it.